File: solutions/go/rest-api/2/rest_api.go
package restapi

// Define the Rest API interface. You should not modify the code in this block.

type User struct {
	Name    string
	Owes    map[string]float64
	OwedBy  map[string]float64
	Balance float64
}

type GetUsersRequest struct {
	Users []string
}

type GetUsersResponse struct {
	Users []User
}

type AddUserRequest struct {
	User string
}

type AddUserResponse struct {
	User User
}

type AddIouRequest struct {
	Lender   string
	Borrower string
	Amount   float64
}

type AddIouResponse struct {
	Users []User
}

type RestApi interface {
	GetUsers(GetUsersRequest) GetUsersResponse
	AddUser(AddUserRequest) AddUserResponse
	AddIou(AddIouRequest) AddIouResponse
}

// Your code goes below here. Implement the RestApi interface.

type Api struct {
	db map[string]User
}

func NewApi(database []User) RestApi {
	result := make(map[string]User)
	for _, u := range database {
		result[u.Name] = u
	}
	return &Api{
		db: result,
	}
}

func (a *Api) GetUsers(req GetUsersRequest) GetUsersResponse {
	var result = make([]User, 0)
	for _, u := range req.Users {
		if user, ok := a.db[u]; ok {
			result = append(result, user)
		}
	}
	return GetUsersResponse{result}
}

func (a *Api) AddUser(req AddUserRequest) AddUserResponse {
	if _, ok := a.db[req.User]; !ok {
		a.db[req.User] = User{
			Name:    req.User,
			Owes:    make(map[string]float64),
			OwedBy:  make(map[string]float64),
			Balance: 0.0,
		}
	}
	return AddUserResponse{a.db[req.User]}
}
func calculateBalance(user *User) {
	user.Balance = 0
	for _, a := range user.OwedBy {
		user.Balance += a
	}
	for _, a := range user.Owes {
		user.Balance -= a
	}
}

func (a *Api) update(req AddIouRequest) (User, User) {
	lender := a.db[req.Lender]
	borrower := a.db[req.Borrower]

	if lenderOwesBack, ok := lender.Owes[req.Borrower]; ok {
		if lenderOwesBack > req.Amount {
			lender.Owes[req.Borrower] = lenderOwesBack - req.Amount
			borrower.OwedBy[req.Lender] = lender.Owes[req.Borrower]
			delete(lender.OwedBy, req.Borrower)
			delete(borrower.Owes, req.Lender)
		} else if lenderOwesBack < req.Amount {

			delete(lender.Owes, req.Borrower)
			delete(borrower.OwedBy, req.Lender)

			lender.OwedBy[req.Borrower] = req.Amount - lenderOwesBack
			borrower.Owes[req.Lender] = req.Amount - lenderOwesBack
		} else {
			delete(lender.Owes, req.Borrower)
			delete(borrower.OwedBy, req.Lender)
		}
	} else if req.Amount > 0 {
		lender.OwedBy[req.Borrower] = req.Amount
		borrower.Owes[req.Lender] = req.Amount
	}
	calculateBalance(&lender)
	calculateBalance(&borrower)

	a.db[req.Lender] = lender
	a.db[req.Borrower] = borrower

	return lender, borrower

}

func (a *Api) AddIou(req AddIouRequest) AddIouResponse {
	lender, borrower := a.update(req)
	users := make([]User, 0)
	if req.Borrower < req.Lender {
		users = append(users, borrower)
		users = append(users, lender)
	} else {
		users = append(users, lender)
		users = append(users, borrower)

	}
	return AddIouResponse{Users: users}

}

File: solutions/csharp/rest-api/2/RestApi.cs
using System.Collections.Concurrent;
using System.Text.Json;
using System.Text.Json.Serialization;

public class User
{
    [JsonPropertyName("name")] public string Name { get; set; }
    [JsonPropertyName("owes")] public SortedDictionary<string, int> Owes { get; set; }

    [JsonPropertyName("owed_by")] public SortedDictionary<string, int> OwedBy { get; set; }

    [JsonPropertyName("balance")] public int Balance { get; set; }
}

public record CreateUserRequest([property: JsonPropertyName("user")] string Name);

public record UserRequest(
    [property: JsonPropertyName("users")] HashSet<string> Users
);

public record IoURequest(
    [property: JsonPropertyName("lender")] string Lender,
    [property: JsonPropertyName("borrower")]
    string Borrower,
    [property: JsonPropertyName("amount")] int Amount
);
public class RestApi
{
    private readonly Dictionary<string, User> _users = new();

    public RestApi(string database)
    {
        var users = JsonSerializer.Deserialize<List<User>>(database) ?? [];
        users.ForEach(s => _users.TryAdd(s.Name, s));
    }

    public string Get(string url, string? payload = null)
    {
        var req = payload == null ? new UserRequest(new()) : JsonSerializer.Deserialize<UserRequest>(payload ?? "");

        var result = _users.Where(s => req.Users.Contains(s.Key))
            .Select(s => s.Value).ToList();
        return JsonSerializer.Serialize(result);
    }

    public string Post(string url, string payload)
    {
        if (url.CompareTo("/iou", StringComparison.OrdinalIgnoreCase) == 0)
        {
            var iou = JsonSerializer.Deserialize<IoURequest>(payload)!;
            UpdateBorrowersAndLenders(iou);
            return JsonSerializer.Serialize(_users
                .Where(s => s.Value.Name == iou.Borrower || s.Value.Name == iou.Lender)
                .Select(s => s.Value)
                .ToList());
        }

        if (url.CompareTo("/add", StringComparison.OrdinalIgnoreCase) == 0)
        {
            var addUser = JsonSerializer.Deserialize<CreateUserRequest>(payload)!;
            var user = new User
            {
                Name = addUser.Name,
                Owes = new(),
                OwedBy = new(),
                Balance = CalculateBalance(addUser.Name)
            };
            _users.TryAdd(user.Name, user);
            return JsonSerializer.Serialize(user);
        }

        return payload;
    }


    private void UpdateBorrowersAndLenders(IoURequest iou)
    {
        var lenderExists = _users.TryGetValue(iou.Lender, out var lender);
        var borrowerExists = _users.TryGetValue(iou.Borrower, out var borrower);
        if (lenderExists && borrowerExists)
        {
            UpdateUsers(lender, borrower, iou.Amount);
            _users[lender.Name] = lender;
            _users[borrower.Name] = borrower;
            _users[lender.Name].Balance = CalculateBalance(lender.Name);
            _users[borrower.Name].Balance = CalculateBalance(borrower.Name);
        }
        
    }

    private void UpdateUsers(User lender, User borrower, int amount)
    {
        if (lender.Owes.TryGetValue(borrower.Name, out var borrowed))
        {
            if (borrowed < amount)
            {
                lender.Owes.Remove(borrower.Name);
                borrower.OwedBy.Remove(lender.Name);
                lender.OwedBy[borrower.Name] = amount - borrowed;
                borrower.Owes[lender.Name] = amount - borrowed;
            }
            else if (borrowed > amount)
            {
                lender.Owes[borrower.Name] = borrowed - amount;
                borrower.OwedBy[lender.Name] = borrowed - amount;
            }
            else
            {
                lender.Owes.Remove(borrower.Name);
                borrower.OwedBy.Remove(lender.Name);
            }
        }
        else
        {
            lender.OwedBy[borrower.Name] = amount;
            borrower.Owes[lender.Name] = amount;
        }
    }


    private int CalculateBalance(string usernmae)
    {
        int balance = 0;
        _ = _users.TryGetValue(usernmae, out var user);
        foreach (var item in user?.OwedBy ?? [])
        {
            balance += item.Value;
        }


        foreach (var item in user?.Owes ?? [])
        {
            balance -= item.Value;
        }

        return balance;
    }
}